src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,185 @@
+        private Node<T> previous;
+
+        Node(T value) {
+            this.value = value;
+        }
+
+        Node(Node<T> previous, Node<T> next, T value) {
+            this.previous = previous;
+            this.next = next;
+            this.value = value;
+        }
+
+    }
+
+    public MyLinkedList() {
+
+    }
+
     @Override
     public boolean add(T value) {
-        return false;
+        addTail(value);
+        return true;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Index out of range");
+        }
+        if (index == size) {
+            addTail(value);
+        } else {
+            addByIndex(value, index);
+        }
     }
 
     @Override
     public boolean addAll(List<T> list) {
-        return false;
+        if (list == null || list.isEmpty()) {
+            return false;
+        }
+        for (T t : list) {
+            add(t);
+        }
+        return true;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> current = getCurrentNode(index);
+        return current.value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        T oldValue;
+        Node<T> current = getCurrentNode(index);
+        oldValue = current.value;
+
+        current.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> current = getCurrentNode(index);
+        Node<T> next = current.next;
+        Node<T> previous = current.previous;
+        final T deletedValue = current.value;
+
+        if (previous == null) {
+            head = next;
+        } else {
+            previous.next = next;
+            current.previous = null;
+        }
+
+        if (next == null) {
+            tail = previous;
+        } else {
+            next.previous = previous;
+            current.next = null;
+        }
+
+        current.value = null;
+        size--;
+        return deletedValue;
     }
 
     @Override
     public boolean remove(T object) {
+        int indexOfObject = findIndexByValue(object);
+        if (indexOfObject != -1) {
+            remove(indexOfObject);
+            return true;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size <= 0;
+    }
+
+    private void checkIndex(int index) {
+        if (!isIndexInRange(index)) {
+            throw new IndexOutOfBoundsException("Index out of range");
+        }
+    }
+
+    private boolean isIndexInRange(int index) {
+        return index >= 0 && index < size;
+    }
+
+    private Node<T> getCurrentNode(int index) {
+        Node<T> current = head;
```

Comment:
let's firstly check which part of the list has an index
If the index is in the first part, then move from the first node to the index, otherwise - from the last node to the index.